ir/src/graph/mod.rs

```diff
@@ -148,40 +172,290 @@ impl AlgebraicGraph {
+    pub fn insert_placeholder_op(&mut self) -> NodeIndex {
+        let index = self.nodes.len();
+        self.nodes.push(Node {
+            op: Operation::Placeholder,
+        });
+        NodeIndex(index)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct PrettyShared<'a> {
+    pub var_count: usize,
+    pub fn_count: usize,
+    // BTreeMap from function index to function id
+    pub fns: BTreeMap<usize, usize>,
+    pub roots: &'a [NodeIndex],
+}
+
+#[derive(Clone)]
+struct PrettyCtx<'a> {
+    pub graph: &'a MirGraph,
+    pub indent: usize,
+    pub nl: &'a str,
+    pub in_block: bool,
+    pub show_var_names: bool,
+    pub shared: Rc<RefCell<PrettyShared<'a>>>,
+}
+
+impl<'a> PrettyCtx<'a> {
+    fn new(graph: &'a MirGraph, roots: &'a [NodeIndex]) -> Self {
+        let shared = Rc::new(RefCell::new(PrettyShared {
+            var_count: 0,
+            fn_count: 0,
+            fns: BTreeMap::new(),
+            roots,
+        }));
+        Self {
+            graph,
+            indent: 0,
+            nl: "\n",
+            in_block: false,
+            shared,
+            show_var_names: true,
+        }
+    }
+
+    fn add_indent(&self, indent: usize) -> Self {
+        Self {
+            indent: self.indent + indent,
+            ..self.clone()
+        }
+    }
+
+    fn with_indent(&self, indent: usize) -> Self {
+        Self {
+            indent,
+            ..self.clone()
+        }
+    }
+
+    fn increment_var_count(&self) -> Self {
+        self.shared.borrow_mut().var_count += 1;
+        self.clone()
+    }
+
+    fn increment_fn_count(&self, node_idx: &NodeIndex) -> Self {
+        let fn_count = self.shared.borrow().fn_count;
+        self.shared.borrow_mut().fns.insert(node_idx.0, fn_count);
+        self.shared.borrow_mut().fn_count += 1;
+        self.clone()
+    }
+
+    fn with_nl(&self, nl: &'a str) -> Self {
+        Self { nl, ..self.clone() }
+    }
+
+    fn with_in_block(&self, in_block: bool) -> Self {
+        Self {
+            in_block,
+            ..self.clone()
+        }
+    }
+
+    fn indent_str(&self) -> String {
+        if self.nl == "\n" {
+            "  ".repeat(self.indent)
+        } else {
+            "".to_string()
+        }
+    }
+
+    fn show_var_names(&self, show_var_names: bool) -> Self {
+        Self {
+            show_var_names,
+            ..self.clone()
+        }
+    }
+}
+
+impl std::fmt::Debug for PrettyCtx<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PrettyCtx")
+            .field("indent", &self.indent)
+            .field("nl", &self.nl)
+            .field("in_block", &self.in_block)
+            .field("var_count", &self.shared.borrow().var_count)
+            .field("fn_count", &self.shared.borrow().fn_count)
+            .finish()
+    }
+}
+
+pub fn pretty(graph: &MirGraph, roots: &[NodeIndex]) -> String {
```

Comment:
Noted, we'll convert our implementation to use miden-formatting instead of our custom implementation once other points discussed here are resolved/implemented.